http-client/src/main/java/com/facebook/airlift/http/client/thrift/ThriftResponseHandler.java

```diff
@@ -0,0 +1,146 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.airlift.http.client.thrift;
+
+import com.facebook.airlift.http.client.HeaderName;
+import com.facebook.airlift.http.client.Request;
+import com.facebook.airlift.http.client.Response;
+import com.facebook.airlift.http.client.ResponseHandler;
+import com.facebook.airlift.http.client.ResponseHandlerUtils;
+import com.facebook.drift.codec.ThriftCodec;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.net.MediaType;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.util.List;
+
+import static com.facebook.airlift.http.client.thrift.ThriftProtocol.BINARY;
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+
+public class ThriftResponseHandler<T>
+        implements ResponseHandler<ThriftResponseHandler.ThriftResponse, RuntimeException>
+{
+    private final ThriftCodec<T> thriftCodec;
+
+    public ThriftResponseHandler(ThriftCodec<T> thriftCodec)
+    {
+        this.thriftCodec = thriftCodec;
+    }
+
+    @Override
+    public ThriftResponse handleException(Request request, Exception exception)
+    {
+        throw ResponseHandlerUtils.propagate(request, exception);
+    }
+
+    @Override
+    public ThriftResponse handle(Request request, Response response)
+    {
+        try {
+            return new ThriftResponse(response.getStatusCode(), response.getStatusMessage(), response.getHeaders(), thriftCodec, response.getInputStream());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public class ThriftResponse
+    {
+        private final int statusCode;
+        private final String statusMessage;
+        private final ListMultimap<HeaderName, String> headers;
+        private final T value;
+        private final IllegalArgumentException exception;
+
+        ThriftResponse(int statusCode, String statusMessage, ListMultimap<HeaderName, String> headers, ThriftCodec<T> thriftCodec, InputStream thriftBytes)
+        {
+            this.statusCode = statusCode;
+            this.statusMessage = statusMessage;
+            this.headers = ImmutableListMultimap.copyOf(headers);
+            T value = null;
+            IllegalArgumentException exception = null;
+            ThriftProtocol thriftProtocol = getThriftProtocol(headers);
```

Comment:
I would recommend moving all the logic up to the `public ThriftResponse handle(Request request, Response response)`, and have a trivial constructor for the `ThriftResponse`. It simplifies the exception handling logic.